datreant/backends/core.py

```diff
@@ -316,3 +313,115 @@ def _close(self):
         self._unlock(self.fd)
         self.fdlock = None
         self._close_fd()
+
+
+class FileSerial(File):
+    """File object base class for serialization formats, such as JSON.
+
+    """
+    def _open_file_r(self):
+        return open(self.filename, 'r')
+
+    def _open_file_w(self):
+        return open(self.filename, 'w')
+
+    @staticmethod
+    def _read(func):
+        """Decorator for opening state file for reading and applying shared
+        lock.
+
+        Applying this decorator to a method will ensure that the file is opened
+        for reading and that a shared lock is obtained before that method is
+        executed. It also ensures that the lock is removed and the file closed
+        after the method returns.
+
+        """
+        @wraps(func)
+        def inner(self, *args, **kwargs):
+            if self.fdlock:
```

Comment:
This is a check for when **this** process already has any kind of lock on the file, either 'shared' or 'exclusive' (which happens for writes). For reading, this process can have either kind of lock already before it does any kind of reading. This check allows a method that itself already has a lock on the file to call another method that would first get a lock. We don't want the called method to change the lock, so this just lets it proceed.